### PR TITLE
feat: Add support for Debian versioning scheme

### PR DESCRIPTION
**Pull request by milux**
- add support for Debian versioning scheme
- fix/enhance dynamic range functionality

### CODE REVIEW

File: src/constraint.rs [hunk 1]
```diff
@@ -93,7 +93,13 @@ impl<V: VersionType> VersionConstraint<V> {
             });
         }
 
-        let (comparator, version) = if constraint_str.starts_with(">=") {
+        // TODO: Move non-vers constraints to their respective implementations,
+        // i.e., implement a from_native() function for each scheme
+        let (comparator, version) = if constraint_str.starts_with("<<") {
+            (Comparator::LessThan, &constraint_str[2..])
+        } else if constraint_str.starts_with(">>") {
```
Comment (tschmidtb51) [inline src/constraint.rs]: Same here with `>>`.

File: src/lib.rs [hunk 2]
```diff
@@ -291,6 +291,34 @@ mod tests {
         assert_eq!(range.constraints().len(), 2);
     }
 
+    #[test]
+    fn test_dynamic_parse_deb() {
+        let range: DynamicVersionRange = "vers:deb/<<1.0".parse().unwrap();
```
Comment (tschmidtb51) [inline src/lib.rs]: This is IMHO not a valid vers as it uses `<<`.

File: src/lib.rs [hunk 3]
```diff
@@ -291,6 +291,34 @@ mod tests {
...
+    }
+
+    #[test]
+    fn test_deb_version_ordering_tilde_and_epoch() {
+        // 1.0~beta < 1.0
+        let range1: DynamicVersionRange = "vers:deb/<<1.0".parse().unwrap();
+        assert!(contains(&range1, "1.0~beta".to_string()).unwrap());
+
+        // 1:1.0 > 2.0 because epoch 1 > 0
+        let range2: DynamicVersionRange = "vers:deb/>>2.0".parse().unwrap();
```
Comment (tschmidtb51) [inline src/lib.rs]: And here.

**Comment by milux** [comment]
@peinjoh Are you the right person to ask for review here?
Please let me know if you have any questions or feel not responsible here.

**Comment by peinjoh** [comment]
> @peinjoh Are you the right person to ask for review here? Please let me know if you have any questions or feel not responsible here.

Hey there, thank you for the submission! I think I'm one of the right persons for the review, but you can tag @csaf-rs/service-dev in the future in case I'm out of office or similar. 

I'll get back to you if questions arise during review, but would like to point you to the ongoing discussion at csaf-rs (https://github.com/csaf-rs/csaf/pull/700) concerning AI usage for contributions, and would kindly ask you to remove your AGENTS.md from this PR for the time being.

**Comment by milux** [comment]
> [...] but would like to point you to the ongoing discussion at csaf-rs concerning AI usage for contributions, and would kindly ask you to remove your AGENTS.md from this PR for the time being.

Sorry, I thought AI stuff would be allowed for quality-improvement-like tooling. Was probably mistaken here. Will remove for the time being. Can you give me a direct link to the discussion?

**Comment by peinjoh** [comment]
I updated my previous comment with a link. Your use case (AI tooling for quality improvement) might be a valueable addition to the discussion there: https://github.com/csaf-rs/csaf/pull/700/changes#r3557310591

**Comment by milux** [comment]
@csaf-rs/service-dev Still awaiting approval. Since this is the first added scheme and I had to do extend the basic structures to some degree for this, adding any more schemes is effectively blocked until this got merged.

**Comment by peinjoh** [comment]
Whoops, that was the wrong button in my IDE, sorry for that. Reviewing now.

File: src/constraint.rs [hunk 4]
```diff
@@ -7,17 +7,92 @@
...
+    /// The vers scheme identifier for this version type (e.g. `"deb"`, `"semver"`).
+    const SCHEME_NAME: &'static str;
+
+    /// Parse a native range string into a fully parsed `VersVersionRange`.
+    ///
+    /// This is the main entry point for converting native syntax into vers ranges.
+    /// The default implementation calls [`Self::from_native`] and wraps
+    /// the result with [`Self::SCHEME_NAME`]. Schemes whose native syntax
+    /// requires special handling can override this directly.
+    fn from_native_string(raw: &str) -> Result<VersVersionRange<Self>, VersError> {
```
Comment (oxisto) [inline src/constraint.rs]: When do we need to call `normalize_and_validate` and when not? Seems the `FromStr` trait of `VersVersonRange` does.

File: src/range/dynamic.rs [hunk 5]
```diff
@@ -27,18 +40,85 @@ use std::str::FromStr;
...
+    /// use vers_rs::range::dynamic::DynamicVersionRange;
+    /// use vers_rs::range::VersionRange;
+    ///
+    /// let range = DynamicVersionRange::parse_native("deb", "<<1.0").unwrap();
+    /// assert_eq!(range.versioning_scheme(), "deb");
+    /// assert!(range.contains("0.9".to_string()).unwrap());
+    /// ```
+    pub fn parse_native(scheme: &str, raw: &str) -> Result<Self, VersError> {
+        let inner = match scheme {
+            "semver" | "npm" => DynamicVersionRangeInner::SemVer(SemVer::from_native_string(raw)?),
```
Comment (oxisto) [inline src/range/dynamic.rs]: I think the "npm" label gets lost here, dont we need to supply that to the version range inner?

File: src/schemes/deb.rs [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+impl NativeVersionConverter for DebVersion {
+    const SCHEME_NAME: &'static str = "deb";
+
+    /// Parse a Debian native constraint string into standard vers constraints.
+    ///
+    /// Debian defines `>>` (strictly greater than) and `<<` (strictly less than)
+    /// as native comparison operators. This method converts them to the
+    /// standard `>` and `<` operators and delegates to the vers parser.
+    fn from_native_constraint(raw: &str) -> Result<VersionConstraint<Self>, VersError> {
+        let converted = raw.replace(">>", ">").replace("<<", "<");
```
Comment (oxisto) [inline src/schemes/deb.rs]: what about `=`, isn't that Debian only?

File: src/range/dynamic.rs [hunk 7]
```diff
@@ -27,18 +40,85 @@ use std::str::FromStr;
...
+/// assert!(npm_range.contains("1.5.0".to_string()).unwrap());
+/// assert!(!npm_range.contains("2.0.0".to_string()).unwrap());
 /// ```
-#[derive(PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
+#[derive(Debug)]
 #[cfg_attr(feature = "wasm", derive(tsify::Tsify))]
 #[cfg_attr(feature = "wasm", tsify(into_wasm_abi, from_wasm_abi))]
-pub enum DynamicVersionRange {
-    /// SemVer-based range (for "semver" and "npm" schemes)
-    SemVer(GenericVersionRange<SemVer>),
```
Comment (oxisto) [inline src/range/dynamic.rs:38]: I think the README still has GenericVersionRange

File: src/schemes/deb.rs [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+}
+
+/// Debian version according to dpkg version format: [epoch:]upstream[-debian]
+///
+/// This implementation follows Debian Policy version comparison rules:
+/// - Epoch numeric (default 0)
+/// - Upstream version vs. Debian revision separated by last '-'
+/// - Tilde '~' sorts before the end and before any other character
+/// - Sequences of digits are compared numerically; non-digits lexicographically
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
```
Comment (oxisto) [inline src/schemes/deb.rs]: I think you need to drop PartialEq and Eq from the derive and implement this yourself, otherwise there might be inconsistencies petween partial ordering and equality because of the Debian revision "fix" in cmp (to set it "0" if its empty)


```rust
impl PartialEq for DebVersion {
      fn eq(&self, other: &Self) -> bool {
          self.cmp(other) == Ordering::Equal
      }
  }
  ```

File: src/schemes/deb.rs (hunk 6 above)
Comment (peinjoh) [reply src/schemes/deb.rs]: I believe this touches https://github.com/csaf-rs/vers-rs/issues/15 and https://github.com/oasis-tcs/csaf/issues/1445 (Issue is on vls, but there might be a side effect due to the vls <> vers similarity).

As long as #15 is not fixed, I believe this code will handle debian version constraints like `=1.0.0` wrong.

Comment (peinjoh) [reply src/schemes/deb.rs]: I believe this also allows too much leeway in the debian specific input. Consider the following, invalid debian versions string: `>>=1.0.0` `<<=1.0.0`. These would get turned into `>=1.0` / `<=1.0` respectivly, turning invalid debian versions into valid vers versions. Also `!=`, `<` and `>` are not valid debian version comparators, and would similarly be turned into valid vers version constraints.

File: src/range/dynamic.rs (hunk 5 above)
Comment (peinjoh) [reply src/range/dynamic.rs]: It does, the semver / npm schemes all come out with semver, as this is the `SCHEME_NAME` provided for both. This breaks roundtrip parsing.

File: src/constraint.rs (hunk 4 above)
Comment (peinjoh) [reply src/constraint.rs]: Same question, I think normalize_and_validate should run irrespective of if the schema was supplied directly or autodetected. Currenlty, it will only run as part of parsing that includes `vers:<scheme>`, and not the newly added parse_native API.

**Comment by peinjoh** [comment]
General comment: I would wager for a bit more test coverage. This does not need to be done in this PR, but could be done as part of #13.

File: src/schemes/deb.rs (hunk 6 above)
Comment (milux) [reply src/schemes/deb.rs]: Already had the subtile feeling that just doing text replacements might not be sufficient here. Thanks for the counter-example pointing that mistake out. Patch incoming...

File: src/range/dynamic.rs (hunk 5 above)
Comment (milux) [reply src/range/dynamic.rs]: Right, impls have to carry that along. Fix on the way.

File: src/constraint.rs (hunk 4 above)
Comment (milux) [reply src/constraint.rs]: I think there was a fundamental decision after that code was initially created. If I recall it correctly, we do not want to normalize anymore **at all**, but instead only validate for correct ordering and raise an error in case a vers string is wrongly ordered. For the time being, I'm going to fix it as you suggested, though...

**Comment by milux** [comment]
I tried to address all review concerns, ready for round 2.

**Comment by milux** [comment]
> General comment: I would wager for a bit more test coverage. This does not need to be done in this PR, but could be done as part of #13.

@peinjoh I added some tests especially targeting the edge cases and bugs pointed out during your review. Do you consider that sufficient or shall I add even more tests? And are you talking about hard, numerical coverage (do we have coverage metrics by our current tooling) or more like perceived coverage of relevant code paths?

**Comment by peinjoh** [comment]
On it now!

File: src/range/dynamic.rs (hunk 5 above)
Comment (peinjoh) [reply src/range/dynamic.rs]: Looking good!

File: src/schemes/deb.rs (hunk 6 above)
Comment (peinjoh) [reply src/schemes/deb.rs]: Looking good, I imagine we'll have to do this "kind" of constraint parsing for every new versioning scheme now 👍

File: src/schemes/deb.rs (hunk 8 above)
Comment (peinjoh) [reply src/schemes/deb.rs]: Looking good, IMO, @oxisto please reopen if you see that this still needs more work done.

File: src/constraint.rs (hunk 4 above)
Comment (peinjoh) [reply src/constraint.rs]: Then we should document that decision (and follow through with it). Do you have some link for this? Otherwise, this fixes my comment of the checks being asymetrical, so im resolving this for now. We should consider adding an issue to implement the decision mentioned above.

**Comment by peinjoh** [comment]
> > General comment: I would wager for a bit more test coverage. This does not need to be done in this PR, but could be done as part of #13.
> 
> @peinjoh I added some tests especially targeting the edge cases and bugs pointed out during your review. Do you consider that sufficient or shall I add even more tests? And are you talking about hard, numerical coverage (do we have coverage metrics by our current tooling) or more like perceived coverage of relevant code paths?

Test coverage looks good IMO. I would be in favor of adding some serde roundtrip tests (specially around the npm/semver code). But as we have few of these tests at all in the library, I would see this as another issue and out of scope of this PR. (I opened #40 for this)

**Review by peinjoh — APPROVED** [review]

**Comment by peinjoh** [comment]
@milux is there anything left to do here? Otherwise, please answer here: https://github.com/csaf-rs/vers-rs/pull/12#discussion_r3657014664 if you have a pointer for the normalization decision mentioned and feel free to merge 👍

**Comment by milux** [comment]
> @milux is there anything left to do here? Otherwise, please answer here: [#12 (comment)](https://github.com/csaf-rs/vers-rs/pull/12#discussion_r3657014664) if you have a pointer for the normalization decision mentioned and feel free to merge 👍

I have requested these links because I only knew about it second hand at that time and online sources still seem to point at normalization instead of a validation-only approach.

**Comment by peinjoh** [comment]
Alright, thanks! Please open an issue with the decision links when you get your hands on them 👍 

And thank your for this PR, looking forward to the next 🚀

**Comment by milux** [comment]
@peinjoh Here we go: https://github.com/package-url/vers-spec/pull/65
Here a reference impl for Python (merge pending) from the same colleague: https://github.com/aboutcode-org/univers/pull/200

**Comment by milux** [comment]
See #41

File: src/constraint.rs (hunk 4 above)
Comment (milux) [reply src/constraint.rs]: See #41
